implementations/rust/ockam/ockam_api/src/nodes/service/relay.rs

```diff
@@ -122,81 +122,89 @@ impl NodeManager {
     /// when the route is unresponsive
     pub async fn create_relay(
-        &self,
+        self: &Arc<Self>,
         ctx: &Context,
-        connection: Connection,
+        addr: &MultiAddr,
+        alias: String,
         at_rust_node: bool,
-        alias: Option<String>,
+        authorized: Option<Identifier>,
+        relay_address: Option<String>,
     ) -> Result<RelayInfo> {
-        let route = connection.route(self.tcp_transport()).await?;
-        let options = RemoteRelayOptions::new();
-
-        let relay = if at_rust_node {
-            if let Some(alias) = alias {
-                RemoteRelay::create_static_without_heartbeats(ctx, route, alias, options).await
-            } else {
-                RemoteRelay::create(ctx, route, options).await
-            }
-        } else if let Some(alias) = alias {
-            RemoteRelay::create_static(ctx, route, alias, options).await
-        } else {
-            RemoteRelay::create(ctx, route, options).await
+        let replacer = RelaySessionReplacer {
+            node_manager: self.clone(),
+            context: Arc::new(ctx.async_try_clone().await?),
+            addr: addr.clone(),
+            at_rust_node,
+            relay_address,
+            connection: None,
+            relay_worker_address: None,
+            authorized,
         };
 
-        match relay {
-            Ok(info) => {
-                let registry_info = info.clone();
-                let registry_remote_address = registry_info.remote_address().to_string();
-                let relay_info = RelayInfo::from(info);
-                self.registry
-                    .relays
-                    .insert(registry_remote_address, registry_info)
-                    .await;
-
-                debug!(
-                    forwarding_route = %relay_info.forwarding_route(),
-                    remote_address = %relay_info.remote_address_ma()?,
-                    "CreateRelay request processed, sending back response"
-                );
-                Ok(relay_info)
-            }
-            Err(err) => {
-                error!(?err, "Failed to create relay");
-                Err(err)
-            }
+        let mut session = Session::new(replacer);
+        let relay_info =
+            MedicHandle::connect(&mut session)
+                .await
+                .map(|outcome| match outcome.kind {
+                    ReplacerOutputKind::Relay(status) => status,
+                    _ => {
+                        panic!("Unexpected outcome: {:?}", outcome);
+                    }
+                })?;
+
+        if self.registry.relays.contains_key(&alias).await {
+            let message = format!("A relay with the name '{alias}' already exists");
+            return Err(ockam_core::Error::new(
+                Origin::Node,
+                Kind::AlreadyExists,
+                message,
+            ));
```

Comment:
Why is the relay created before checking if it already exists? In the `create_inlet` function it's doing the check first.

Reply:
no particular reason, it makes sense to delay the creation after full validation